File: sdk/python/src/fullstack_api/storage.py
"""Token storage implementations."""

import json
import os
from abc import ABC, abstractmethod
from pathlib import Path
from typing import Optional

from .types import AuthTokens


class TokenStorage(ABC):
    """Abstract base class for token storage."""

    @abstractmethod
    def get_access_token(self) -> Optional[str]:
        """Get stored access token."""
        pass

    @abstractmethod
    def get_refresh_token(self) -> Optional[str]:
        """Get stored refresh token."""
        pass

    @abstractmethod
    def set_tokens(self, tokens: AuthTokens) -> None:
        """Store tokens."""
        pass

    @abstractmethod
    def clear_tokens(self) -> None:
        """Clear stored tokens."""
        pass


class MemoryTokenStorage(TokenStorage):
    """In-memory token storage."""

    def __init__(self):
        """Initialize storage."""
        self._access_token: Optional[str] = None
        self._refresh_token: Optional[str] = None

    def get_access_token(self) -> Optional[str]:
        """Get stored access token."""
        return self._access_token

    def get_refresh_token(self) -> Optional[str]:
        """Get stored refresh token."""
        return self._refresh_token

    def set_tokens(self, tokens: AuthTokens) -> None:
        """Store tokens."""
        self._access_token = tokens.access_token
        self._refresh_token = tokens.refresh_token

    def clear_tokens(self) -> None:
        """Clear stored tokens."""
        self._access_token = None
        self._refresh_token = None


class FileTokenStorage(TokenStorage):
    """File-based token storage."""

    def __init__(self, file_path: Optional[str] = None):
        """Initialize storage.
        
        Args:
            file_path: Path to token file. Defaults to ~/.fullstack/tokens.json
        """
        if file_path is None:
            home = Path.home()
            self.file_path = home / ".fullstack" / "tokens.json"
        else:
            self.file_path = Path(file_path)
        
        # Create directory if it doesn't exist
        self.file_path.parent.mkdir(parents=True, exist_ok=True)

    def _read_tokens(self) -> dict:
        """Read tokens from file."""
        if not self.file_path.exists():
            return {}
        
        try:
            with open(self.file_path, "r") as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError):
            return {}

    def _write_tokens(self, tokens: dict) -> None:
        """Write tokens to file."""
        with open(self.file_path, "w") as f:
            json.dump(tokens, f)
        
        # Set restrictive permissions (owner read/write only)
        os.chmod(self.file_path, 0o600)

    def get_access_token(self) -> Optional[str]:
        """Get stored access token."""
        tokens = self._read_tokens()
        return tokens.get("access_token")

    def get_refresh_token(self) -> Optional[str]:
        """Get stored refresh token."""
        tokens = self._read_tokens()
        return tokens.get("refresh_token")

    def set_tokens(self, tokens: AuthTokens) -> None:
        """Store tokens."""
        self._write_tokens({
            "access_token": tokens.access_token,
            "refresh_token": tokens.refresh_token,
            "token_type": tokens.token_type,
        })

    def clear_tokens(self) -> None:
        """Clear stored tokens."""
        if self.file_path.exists():
            self.file_path.unlink()

File: sdk/typescript/src/storage.ts
/**
 * Token storage implementations
 */

import { AuthTokens, TokenStorage } from './types';

/**
 * In-memory token storage (suitable for Node.js applications)
 */
export class MemoryTokenStorage implements TokenStorage {
  private accessToken: string | null = null;
  private refreshToken: string | null = null;

  getAccessToken(): string | null {
    return this.accessToken;
  }

  getRefreshToken(): string | null {
    return this.refreshToken;
  }

  setTokens(tokens: AuthTokens): void {
    this.accessToken = tokens.access_token;
    this.refreshToken = tokens.refresh_token;
  }

  clearTokens(): void {
    this.accessToken = null;
    this.refreshToken = null;
  }
}

/**
 * LocalStorage token storage (for browser applications)
 */
export class LocalStorageTokenStorage implements TokenStorage {
  private readonly accessTokenKey = 'fullstack_access_token';
  private readonly refreshTokenKey = 'fullstack_refresh_token';

  getAccessToken(): string | null {
    if (typeof window === 'undefined') return null;
    return localStorage.getItem(this.accessTokenKey);
  }

  getRefreshToken(): string | null {
    if (typeof window === 'undefined') return null;
    return localStorage.getItem(this.refreshTokenKey);
  }

  setTokens(tokens: AuthTokens): void {
    if (typeof window === 'undefined') return;
    localStorage.setItem(this.accessTokenKey, tokens.access_token);
    localStorage.setItem(this.refreshTokenKey, tokens.refresh_token);
  }

  clearTokens(): void {
    if (typeof window === 'undefined') return;
    localStorage.removeItem(this.accessTokenKey);
    localStorage.removeItem(this.refreshTokenKey);
  }
}

/**
 * SessionStorage token storage (for browser applications with enhanced security)
 */
export class SessionStorageTokenStorage implements TokenStorage {
  private readonly accessTokenKey = 'fullstack_access_token';
  private readonly refreshTokenKey = 'fullstack_refresh_token';

  getAccessToken(): string | null {
    if (typeof window === 'undefined') return null;
    return sessionStorage.getItem(this.accessTokenKey);
  }

  getRefreshToken(): string | null {
    if (typeof window === 'undefined') return null;
    return sessionStorage.getItem(this.refreshTokenKey);
  }

  setTokens(tokens: AuthTokens): void {
    if (typeof window === 'undefined') return;
    sessionStorage.setItem(this.accessTokenKey, tokens.access_token);
    sessionStorage.setItem(this.refreshTokenKey, tokens.refresh_token);
  }

  clearTokens(): void {
    if (typeof window === 'undefined') return;
    sessionStorage.removeItem(this.accessTokenKey);
    sessionStorage.removeItem(this.refreshTokenKey);
  }
}
